File: backend/services/options.py
"""REAL options data via alpaca-py with graceful mock fallback.

- Expirations: TradingClient.get_option_contracts (paginated) → distinct dates.
- Chain: OptionHistoricalDataClient.get_option_chain → mapped contract list.
- Flow: aggregated from the real chain for the relevant expiration.

Every call falls back to services.mock.* on failure / empty so the UI keeps working.
"""
from __future__ import annotations

import re
from datetime import date, datetime, timedelta, timezone
from typing import Any

from config import logger, settings
from services import mock

# ---- Lazy SDK import --------------------------------------------------------
_sdk_ok = False
try:
    from alpaca.trading.client import TradingClient
    from alpaca.trading.requests import GetOptionContractsRequest
    from alpaca.data.historical.option import OptionHistoricalDataClient
    from alpaca.data.requests import OptionChainRequest
    from alpaca.data.enums import OptionsFeed

    _sdk_ok = True
except Exception as exc:  # pragma: no cover
    logger.warning("alpaca options SDK not importable (%s) — options mock-only.", exc)
    _sdk_ok = False

_trading_client = None
_option_client = None

# OCC option symbol: ROOT(1-6) + YYMMDD(6) + C|P + strike*1000 padded to 8.
OCC_RE = re.compile(r"^([A-Z]{1,6})(\d{6})([CP])(\d{8})$")


def _trading():
    global _trading_client
    if not (_sdk_ok and settings.alpaca_configured):
        return None
    if _trading_client is None:
        _trading_client = TradingClient(
            settings.alpaca_api_key, settings.alpaca_secret_key,
            paper=settings.alpaca_paper_trade,
        )
    return _trading_client


def _option_data():
    global _option_client
    if not (_sdk_ok and settings.alpaca_configured):
        return None
    if _option_client is None:
        _option_client = OptionHistoricalDataClient(
            settings.alpaca_api_key, settings.alpaca_secret_key
        )
    return _option_client


def _feed():
    val = (settings.alpaca_options_feed or "indicative").lower()
    try:
        return OptionsFeed.OPRA if val == "opra" else OptionsFeed.INDICATIVE
    except Exception:
        return None


# ---- OCC parsing ------------------------------------------------------------
def parse_occ(occ: str) -> dict[str, Any] | None:
    m = OCC_RE.match(occ.strip().upper())
    if not m:
        return None
    root, ymd, cp, strike_raw = m.groups()
    try:
        exp = datetime.strptime(ymd, "%y%m%d").date().isoformat()
    except ValueError:
        return None
    return {
        "root": root,
        "expiration": exp,
        "type": "call" if cp == "C" else "put",
        "strike": int(strike_raw) / 1000.0,
    }


def is_occ_symbol(symbol: str) -> bool:
    return bool(OCC_RE.match((symbol or "").strip().upper()))


def _is_monthly(d: date) -> bool:
    # Monthly = the 3rd Friday of its month.
    return d.weekday() == 4 and 15 <= d.day <= 21


# ---- Expirations ------------------------------------------------------------
def get_option_expirations(symbol: str) -> list[str]:
    client = _trading()
    if client is None:
        return mock.mock_expirations(symbol)
    try:
        today = datetime.now(timezone.utc).date()
        seen: set[str] = set()
        page_token = None
        pages = 0
        while pages < 10:
            req = GetOptionContractsRequest(
                underlying_symbols=[symbol],
                expiration_date_gte=today,
                limit=10000,
                page_token=page_token,
            )
            resp = client.get_option_contracts(req)
            contracts = getattr(resp, "option_contracts", None) or []
            for c in contracts:
                exp = getattr(c, "expiration_date", None)
                if exp:
                    seen.add(exp.isoformat() if hasattr(exp, "isoformat") else str(exp))
            page_token = getattr(resp, "next_page_token", None)
            pages += 1
            if not page_token:
                break
        out = sorted(seen)
        return out or mock.mock_expirations(symbol)
    except Exception as exc:
        logger.warning("get_option_expirations failed for %s (%s) — mock.", symbol, exc)
        return mock.mock_expirations(symbol)


def expirations_with_type(symbol: str) -> list[dict[str, Any]]:
    out = []
    for e in get_option_expirations(symbol):
        try:
            d = datetime.fromisoformat(e).date()
            t = "monthly" if _is_monthly(d) else "weekly"
        except Exception:
            t = "weekly"
        out.append({"date": e, "type": t})
    return out


# ---- Chain ------------------------------------------------------------------
def _f(x) -> float | None:
    try:
        return float(x) if x is not None else None
    except (TypeError, ValueError):
        return None


def _snapshot_to_contract(occ: str, snap: Any) -> dict[str, Any] | None:
    parsed = parse_occ(occ)
    if not parsed:
        return None
    quote = getattr(snap, "latest_quote", None)
    trade = getattr(snap, "latest_trade", None)
    greeks = getattr(snap, "greeks", None)
    bid = _f(getattr(quote, "bid_price", None)) if quote else None
    ask = _f(getattr(quote, "ask_price", None)) if quote else None
    last = _f(getattr(trade, "price", None)) if trade else None
    bid_size = _f(getattr(quote, "bid_size", None)) if quote else None
    ask_size = _f(getattr(quote, "ask_size", None)) if quote else None
    volume = int(getattr(trade, "size", 0) or 0) if trade else 0
    return {
        "symbol": occ,
        "strike": parsed["strike"],
        "type": parsed["type"],
        "expiration": parsed["expiration"],
        "bid": bid or 0.0,
        "ask": ask or 0.0,
        "last": last if last is not None else round(((bid or 0.0) + (ask or 0.0)) / 2, 2),
        "bid_size": bid_size or 0.0,
        "ask_size": ask_size or 0.0,
        "volume": volume,
        "open_interest": int(getattr(snap, "open_interest", 0) or 0),
        "implied_volatility": round(_f(getattr(snap, "implied_volatility", None)) or 0.0, 6),
        "delta": round(_f(getattr(greeks, "delta", None)) or 0.0, 6) if greeks else 0.0,
        "gamma": round(_f(getattr(greeks, "gamma", None)) or 0.0, 6) if greeks else 0.0,
        "theta": round(_f(getattr(greeks, "theta", None)) or 0.0, 6) if greeks else 0.0,
        "vega": round(_f(getattr(greeks, "vega", None)) or 0.0, 6) if greeks else 0.0,
    }


def get_option_chain(symbol: str, expiration: str | None, opt_type: str) -> list[dict[str, Any]]:
    client = _option_data()
    if client is None:
        return mock.mock_option_chain(symbol, expiration, opt_type)
    try:
        exp = expiration
        if not exp:
            exps = get_option_expirations(symbol)
            exp = exps[0] if exps else None
        kwargs: dict[str, Any] = {"underlying_symbol": symbol}
        feed = _feed()
        if feed is not None:
            kwargs["feed"] = feed
        if exp:
            kwargs["expiration_date"] = datetime.fromisoformat(exp).date()
        req = OptionChainRequest(**kwargs)
        chain = client.get_option_chain(req)
        out: list[dict[str, Any]] = []
        for occ, snap in (chain or {}).items():
            contract = _snapshot_to_contract(occ, snap)
            if contract:
                out.append(contract)
        if opt_type in ("call", "put"):
            out = [c for c in out if c["type"] == opt_type]
        out.sort(key=lambda c: (c["expiration"], c["type"], c["strike"]))
        if not out:
            logger.warning("get_option_chain empty for %s %s — mock.", symbol, exp)
            return mock.mock_option_chain(symbol, expiration, opt_type)
        return out
    except Exception as exc:
        logger.warning("get_option_chain failed for %s (%s) — mock.", symbol, exc)
        return mock.mock_option_chain(symbol, expiration, opt_type)


# ---- Flow -------------------------------------------------------------------
def _nearest_expiration(symbol: str, period: str) -> str | None:
    exps = get_option_expirations(symbol)
    if not exps:
        return None
    if period == "weekly":
        for e in exps:
            try:
                if not _is_monthly(datetime.fromisoformat(e).date()):
                    return e
            except Exception:
                continue
    return exps[0]


def get_option_flow(symbol: str, period: str) -> dict[str, Any]:
    client = _option_data()
    if client is None:
        return mock.mock_option_flow(symbol, period)
    try:
        exp = _nearest_expiration(symbol, period)
        chain = get_option_chain(symbol, exp, "all")
        # If the chain came back mock-shaped (no greeks at all), still aggregate it.
        if not chain:
            return mock.mock_option_flow(symbol, period)

        calls = [c for c in chain if c["type"] == "call"]
        puts = [c for c in chain if c["type"] == "put"]

        def agg(rows):
            out = []
            for r in rows:
                mid = ((r.get("bid") or 0.0) + (r.get("ask") or 0.0)) / 2
                out.append({
                    "strike": r["strike"],
                    "volume": r["volume"],
                    "open_interest": r["open_interest"],
                    "premium": round(r["volume"] * mid * 100, 2),
                    "iv": r["implied_volatility"],
                })
            return out

        total_call_vol = sum(c["volume"] for c in calls)
        total_put_vol = sum(p["volume"] for p in puts)
        unusual = []
        for r in chain:
            oi = max(1, r["open_interest"])
            if r["volume"] > 2 * oi:
                mid = ((r.get("bid") or 0.0) + (r.get("ask") or 0.0)) / 2
                unusual.append({
                    "strike": r["strike"],
                    "type": r["type"],
                    "volume": r["volume"],
                    "oi": r["open_interest"],
                    "vol_oi_ratio": round(r["volume"] / oi, 2),
                    "premium": round(r["volume"] * mid * 100, 2),
                })
        return {
            "expiration": exp or (chain[0]["expiration"] if chain else None),
            "calls": agg(calls),
            "puts": agg(puts),
            "put_call_ratio": round(total_put_vol / max(1, total_call_vol), 4),
            "total_call_volume": total_call_vol,
            "total_put_volume": total_put_vol,
            "unusual": sorted(unusual, key=lambda x: -x["vol_oi_ratio"])[:10],
        }
    except Exception as exc:
        logger.warning("get_option_flow failed for %s (%s) — mock.", symbol, exc)
        return mock.mock_option_flow(symbol, period)

File: backend-rs/src/options.rs
//! REAL options data via Alpaca REST. NO MOCK FALLBACK.
//! Trading API: /v2/options/contracts. Data API: /v1beta1/options/snapshots/{underlying}.

use crate::error::{ApiError, ApiResult};
use crate::state::AppState;
use chrono::{Datelike, NaiveDate, Utc};
use serde_json::{json, Value};

const DATA_BASE: &str = "https://data.alpaca.markets";

fn fopt(v: &Value) -> Option<f64> {
    match v {
        Value::Number(n) => n.as_f64(),
        Value::String(s) => s.parse().ok(),
        _ => None,
    }
}

fn parse_occ(occ: &str) -> Option<Value> {
    let s = occ.trim().to_uppercase();
    if !crate::alpaca::is_occ_symbol(&s) {
        return None;
    }
    let n = s.len();
    let root = &s[..n - 15];
    let ymd = &s[n - 15..n - 9];
    let cp = &s[n - 9..n - 8];
    let strike_raw = &s[n - 8..];
    let exp = NaiveDate::parse_from_str(ymd, "%y%m%d").ok()?;
    let strike: f64 = strike_raw.parse::<f64>().ok()? / 1000.0;
    Some(json!({
        "root": root,
        "expiration": exp.format("%Y-%m-%d").to_string(),
        "type": if cp == "C" { "call" } else { "put" },
        "strike": strike,
    }))
}

fn is_monthly(d: NaiveDate) -> bool {
    d.weekday() == chrono::Weekday::Fri && d.day() >= 15 && d.day() <= 21
}

fn feed(state: &AppState) -> &'static str {
    if state.settings.alpaca_options_feed.to_lowercase() == "opra" {
        "opra"
    } else {
        "indicative"
    }
}

async fn get_trading(state: &AppState, path: &str, query: &[(&str, String)]) -> ApiResult<Value> {
    if !state.settings.alpaca_configured() {
        return Err(ApiError::dependency("Alpaca credentials not configured."));
    }
    let url = format!("{}/v2/{}", state.settings.alpaca_trading_base(), path);
    let resp = state
        .alpaca
        .http
        .get(&url)
        .header("APCA-API-KEY-ID", &state.settings.alpaca_api_key)
        .header("APCA-API-SECRET-KEY", &state.settings.alpaca_secret_key)
        .query(query)
        .send()
        .await
        .map_err(|e| ApiError::upstream(format!("Alpaca options request failed: {e}")))?;
    let status = resp.status();
    let text = resp.text().await.unwrap_or_default();
    if !status.is_success() {
        return Err(ApiError::upstream(format!("Alpaca options {status}: {text}")));
    }
    serde_json::from_str(&text).map_err(|e| ApiError::upstream(format!("options bad JSON: {e}")))
}

async fn get_data(state: &AppState, url: &str, query: &[(&str, String)]) -> ApiResult<Value> {
    if !state.settings.alpaca_configured() {
        return Err(ApiError::dependency("Alpaca credentials not configured."));
    }
    let resp = state
        .alpaca
        .http
        .get(url)
        .header("APCA-API-KEY-ID", &state.settings.alpaca_api_key)
        .header("APCA-API-SECRET-KEY", &state.settings.alpaca_secret_key)
        .query(query)
        .send()
        .await
        .map_err(|e| ApiError::upstream(format!("Alpaca options data request failed: {e}")))?;
    let status = resp.status();
    let text = resp.text().await.unwrap_or_default();
    if !status.is_success() {
        return Err(ApiError::upstream(format!("Alpaca options data {status}: {text}")));
    }
    serde_json::from_str(&text).map_err(|e| ApiError::upstream(format!("options data bad JSON: {e}")))
}

pub async fn get_option_expirations(state: &AppState, symbol: &str) -> ApiResult<Vec<String>> {
    let today = Utc::now().date_naive().format("%Y-%m-%d").to_string();
    let mut seen = std::collections::BTreeSet::new();
    let mut page_token: Option<String> = None;
    let mut pages = 0;
    while pages < 10 {
        let mut q = vec![
            ("underlying_symbols", symbol.to_string()),
            ("expiration_date_gte", today.clone()),
            ("limit", "10000".to_string()),
        ];
        if let Some(pt) = &page_token {
            q.push(("page_token", pt.clone()));
        }
        let resp = get_trading(state, "options/contracts", &q).await?;
        let empty = vec![];
        let contracts = resp["option_contracts"].as_array().unwrap_or(&empty);
        for c in contracts {
            if let Some(exp) = c["expiration_date"].as_str() {
                seen.insert(exp.to_string());
            }
        }
        page_token = resp["next_page_token"].as_str().map(|s| s.to_string());
        pages += 1;
        if page_token.is_none() {
            break;
        }
    }
    let out: Vec<String> = seen.into_iter().collect();
    if out.is_empty() {
        return Err(ApiError::upstream(format!(
            "No option expirations available for {symbol}."
        )));
    }
    Ok(out)
}

pub async fn expirations_with_type(state: &AppState, symbol: &str) -> ApiResult<Value> {
    let exps = get_option_expirations(state, symbol).await?;
    let out: Vec<Value> = exps
        .iter()
        .map(|e| {
            let t = NaiveDate::parse_from_str(e, "%Y-%m-%d")
                .map(|d| if is_monthly(d) { "monthly" } else { "weekly" })
                .unwrap_or("weekly");
            json!({"date": e, "type": t})
        })
        .collect();
    Ok(Value::Array(out))
}

fn snapshot_to_contract(occ: &str, snap: &Value) -> Option<Value> {
    let parsed = parse_occ(occ)?;
    let quote = &snap["latestQuote"];
    let trade = &snap["latestTrade"];
    let greeks = &snap["greeks"];
    let bid = fopt(&quote["bp"]).unwrap_or(0.0);
    let ask = fopt(&quote["ap"]).unwrap_or(0.0);
    let last = fopt(&trade["p"]);
    let bid_size = fopt(&quote["bs"]).unwrap_or(0.0);
    let ask_size = fopt(&quote["as"]).unwrap_or(0.0);
    let volume = trade["s"].as_i64().unwrap_or(0);
    let r6 = |x: f64| (x * 1_000_000.0).round() / 1_000_000.0;
    Some(json!({
        "symbol": occ,
        "strike": parsed["strike"],
        "type": parsed["type"],
        "expiration": parsed["expiration"],
        "bid": bid,
        "ask": ask,
        "last": last.unwrap_or(((bid + ask) / 2.0 * 100.0).round() / 100.0),
        "bid_size": bid_size,
        "ask_size": ask_size,
        "volume": volume,
        "open_interest": snap["openInterest"].as_i64().unwrap_or(0),
        "implied_volatility": r6(fopt(&snap["impliedVolatility"]).unwrap_or(0.0)),
        "delta": if greeks.is_null() { 0.0 } else { r6(fopt(&greeks["delta"]).unwrap_or(0.0)) },
        "gamma": if greeks.is_null() { 0.0 } else { r6(fopt(&greeks["gamma"]).unwrap_or(0.0)) },
        "theta": if greeks.is_null() { 0.0 } else { r6(fopt(&greeks["theta"]).unwrap_or(0.0)) },
        "vega": if greeks.is_null() { 0.0 } else { r6(fopt(&greeks["vega"]).unwrap_or(0.0)) },
    }))
}

pub async fn get_option_chain(
    state: &AppState,
    symbol: &str,
    expiration: Option<&str>,
    opt_type: &str,
) -> ApiResult<Vec<Value>> {
    let exp = match expiration {
        Some(e) => Some(e.to_string()),
        None => get_option_expirations(state, symbol).await?.into_iter().next(),
    };
    let url = format!("{}/v1beta1/options/snapshots/{}", DATA_BASE, symbol);
    let mut q = vec![("feed", feed(state).to_string()), ("limit", "1000".to_string())];
    if let Some(e) = &exp {
        q.push(("expiration_date", e.clone()));
    }
    // paginate snapshots
    let mut out: Vec<Value> = vec![];
    let mut page_token: Option<String> = None;
    loop {
        let mut qq = q.clone();
        if let Some(pt) = &page_token {
            qq.push(("page_token", pt.clone()));
        }
        let resp = get_data(state, &url, &qq).await?;
        if let Some(map) = resp["snapshots"].as_object() {
            for (occ, snap) in map {
                if let Some(c) = snapshot_to_contract(occ, snap) {
                    out.push(c);
                }
            }
        }
        page_token = resp["next_page_token"].as_str().map(|s| s.to_string());
        if page_token.is_none() {
            break;
        }
    }
    if opt_type == "call" || opt_type == "put" {
        out.retain(|c| c["type"].as_str() == Some(opt_type));
    }
    out.sort_by(|a, b| {
        let ka = (a["expiration"].as_str().unwrap_or(""), a["type"].as_str().unwrap_or(""), fopt(&a["strike"]).unwrap_or(0.0));
        let kb = (b["expiration"].as_str().unwrap_or(""), b["type"].as_str().unwrap_or(""), fopt(&b["strike"]).unwrap_or(0.0));
        ka.0.cmp(kb.0).then(ka.1.cmp(kb.1)).then(ka.2.partial_cmp(&kb.2).unwrap_or(std::cmp::Ordering::Equal))
    });
    if out.is_empty() {
        return Err(ApiError::upstream(format!(
            "Empty option chain for {symbol} {}.",
            exp.unwrap_or_else(|| "(nearest)".to_string())
        )));
    }
    Ok(out)
}

pub async fn chain_json(
    state: &AppState,
    symbol: &str,
    expiration: Option<&str>,
    opt_type: &str,
) -> ApiResult<Value> {
    Ok(Value::Array(get_option_chain(state, symbol, expiration, opt_type).await?))
}

async fn underlying_price(state: &AppState, symbol: &str) -> ApiResult<f64> {
    let snap = state.alpaca.get_snapshot(symbol).await?;
    let price = fopt(&snap["price"]).unwrap_or(0.0);
    if price <= 0.0 {
        return Err(ApiError::upstream(format!("No underlying price for {symbol}.")));
    }
    Ok(price)
}

async fn nearest_weekly(state: &AppState, symbol: &str) -> ApiResult<Option<String>> {
    let exps = expirations_with_type(state, symbol).await?;
    let arr = exps.as_array().cloned().unwrap_or_default();
    let mut weeklies: Vec<String> = arr
        .iter()
        .filter(|e| e["type"].as_str() == Some("weekly"))
        .filter_map(|e| e["date"].as_str().map(|s| s.to_string()))
        .collect();
    weeklies.sort();
    if let Some(w) = weeklies.into_iter().next() {
        return Ok(Some(w));
    }
    let mut all: Vec<String> = arr.iter().filter_map(|e| e["date"].as_str().map(|s| s.to_string())).collect();
    all.sort();
    Ok(all.into_iter().next())
}

async fn resolve_expiry(state: &AppState, symbol: &str, expiry: Option<&str>) -> ApiResult<String> {
    let spec = expiry.unwrap_or("nearest_weekly").trim();
    if spec.is_empty() || spec.to_lowercase() == "nearest_weekly" {
        return nearest_weekly(state, symbol)
            .await?
            .ok_or_else(|| ApiError::upstream(format!("No weekly expiration for {symbol}.")));
    }
    if NaiveDate::parse_from_str(spec, "%Y-%m-%d").is_ok() {
        Ok(spec.to_string())
    } else {
        Err(ApiError::bad_request(format!("Bad expiry '{spec}'.")))
    }
}

fn moneyness(strike: f64, under: f64, right: &str) -> &'static str {
    if (strike - under).abs() < 1e-9 {
        return "ATM";
    }
    if right == "call" {
        if strike < under {
            "ITM"
        } else {
            "OTM"
        }
    } else if strike > under {
        "ITM"
    } else {
        "OTM"
    }
}

fn distance_pct(strike: f64, under: f64) -> f64 {
    if under != 0.0 {
        ((strike - under) / under * 100.0 * 10000.0).round() / 10000.0
    } else {
        0.0
    }
}

pub async fn select_contracts(
    state: &AppState,
    symbol: &str,
    right: &str,
    expiry: Option<&str>,
    moneyness_filter: Option<&str>,
    count: usize,
) -> ApiResult<Value> {
    let right = right.to_lowercase();
    if right != "call" && right != "put" {
        return Err(ApiError::bad_request("right must be call or put"));
    }
    let count = count.clamp(1, 50);
    let exp = resolve_expiry(state, symbol, expiry).await?;
    let under = underlying_price(state, symbol).await?;
    let mut chain = get_option_chain(state, symbol, Some(&exp), &right).await?;
    chain.sort_by(|a, b| fopt(&a["strike"]).unwrap_or(0.0).partial_cmp(&fopt(&b["strike"]).unwrap_or(0.0)).unwrap());
    let strikes: Vec<f64> = chain.iter().map(|c| fopt(&c["strike"]).unwrap_or(0.0)).collect();
    let atm_idx = strikes
        .iter()
        .enumerate()
        .min_by(|(_, a), (_, b)| (**a - under).abs().partial_cmp(&(**b - under).abs()).unwrap())
        .map(|(i, _)| i)
        .unwrap_or(0);

    let annotated: Vec<Value> = chain
        .iter()
        .map(|c| {
            let strike = fopt(&c["strike"]).unwrap_or(0.0);
            let bid = fopt(&c["bid"]).unwrap_or(0.0);
            let ask = fopt(&c["ask"]).unwrap_or(0.0);
            let mid = if bid != 0.0 || ask != 0.0 {
                ((bid + ask) / 2.0 * 10000.0).round() / 10000.0
            } else {
                fopt(&c["last"]).unwrap_or(0.0)
            };
            json!({
                "occ_symbol": c["symbol"],
                "strike": strike,
                "right": right,
                "expiration": c["expiration"],
                "bid": bid,
                "ask": ask,
                "mid": mid,
                "last": fopt(&c["last"]).unwrap_or(0.0),
                "implied_volatility": fopt(&c["implied_volatility"]).unwrap_or(0.0),
                "delta": fopt(&c["delta"]).unwrap_or(0.0),
                "gamma": fopt(&c["gamma"]).unwrap_or(0.0),
                "theta": fopt(&c["theta"]).unwrap_or(0.0),
                "vega": fopt(&c["vega"]).unwrap_or(0.0),
                "open_interest": c["open_interest"].as_i64().unwrap_or(0),
                "volume": c["volume"].as_i64().unwrap_or(0),
                "moneyness": moneyness(strike, under, &right),
                "distance_pct": distance_pct(strike, under),
            })
        })
        .collect();

    let selected: Vec<Value> = if let Some(mny) = moneyness_filter {
        let mny = mny.to_uppercase();
        if mny == "ATM" {
            let lo = atm_idx.saturating_sub(count / 2);
            let mut hi = (lo + count).min(annotated.len());
            let lo = hi.saturating_sub(count);
            hi = hi.max(lo);
            annotated[lo..hi].to_vec()
        } else {
            let mut filtered: Vec<Value> = annotated
                .iter()
                .filter(|c| c["moneyness"].as_str() == Some(mny.as_str()))
                .cloned()
                .collect();
            filtered.sort_by(|a, b| (fopt(&a["strike"]).unwrap_or(0.0) - under).abs().partial_cmp(&(fopt(&b["strike"]).unwrap_or(0.0) - under).abs()).unwrap());
            filtered.truncate(count);
            filtered.sort_by(|a, b| fopt(&a["strike"]).unwrap_or(0.0).partial_cmp(&fopt(&b["strike"]).unwrap_or(0.0)).unwrap());
            filtered
        }
    } else {
        let lo = atm_idx.saturating_sub(count / 2);
        let mut hi = lo + count;
        let lo = if hi > annotated.len() {
            hi = annotated.len();
            hi.saturating_sub(count)
        } else {
            lo
        };
        annotated[lo..hi].to_vec()
    };

    Ok(json!({
        "symbol": symbol.to_uppercase(),
        "underlying_price": (under * 10000.0).round() / 10000.0,
        "expiration": exp,
        "right": right,
        "contracts": selected,
    }))
}

pub async fn get_option_flow(state: &AppState, symbol: &str, period: &str) -> ApiResult<Value> {
    let exps = get_option_expirations(state, symbol).await?;
    let exp = if period == "weekly" {
        exps.iter()
            .find(|e| NaiveDate::parse_from_str(e, "%Y-%m-%d").map(|d| !is_monthly(d)).unwrap_or(false))
            .or_else(|| exps.first())
            .cloned()
    } else {
        exps.first().cloned()
    };
    let chain = get_option_chain(state, symbol, exp.as_deref(), "all").await?;
    let calls: Vec<&Value> = chain.iter().filter(|c| c["type"].as_str() == Some("call")).collect();
    let puts: Vec<&Value> = chain.iter().filter(|c| c["type"].as_str() == Some("put")).collect();

    let agg = |rows: &[&Value]| -> Vec<Value> {
        rows.iter()
            .map(|r| {
                let mid = (fopt(&r["bid"]).unwrap_or(0.0) + fopt(&r["ask"]).unwrap_or(0.0)) / 2.0;
                let vol = r["volume"].as_i64().unwrap_or(0);
                json!({
                    "strike": r["strike"],
                    "volume": vol,
                    "open_interest": r["open_interest"],
                    "premium": ((vol as f64 * mid * 100.0) * 100.0).round() / 100.0,
                    "iv": r["implied_volatility"],
                })
            })
            .collect()
    };
    let total_call_vol: i64 = calls.iter().map(|c| c["volume"].as_i64().unwrap_or(0)).sum();
    let total_put_vol: i64 = puts.iter().map(|p| p["volume"].as_i64().unwrap_or(0)).sum();
    let mut unusual = vec![];
    for r in &chain {
        let oi = r["open_interest"].as_i64().unwrap_or(0).max(1);
        let vol = r["volume"].as_i64().unwrap_or(0);
        if vol > 2 * oi {
            let mid = (fopt(&r["bid"]).unwrap_or(0.0) + fopt(&r["ask"]).unwrap_or(0.0)) / 2.0;
            unusual.push(json!({
                "strike": r["strike"],
                "type": r["type"],
                "volume": vol,
                "oi": r["open_interest"],
                "vol_oi_ratio": ((vol as f64 / oi as f64) * 100.0).round() / 100.0,
                "premium": ((vol as f64 * mid * 100.0) * 100.0).round() / 100.0,
            }));
        }
    }
    unusual.sort_by(|a, b| fopt(&b["vol_oi_ratio"]).unwrap_or(0.0).partial_cmp(&fopt(&a["vol_oi_ratio"]).unwrap_or(0.0)).unwrap());
    unusual.truncate(10);

    Ok(json!({
        "expiration": exp.or_else(|| chain.first().and_then(|c| c["expiration"].as_str().map(|s| s.to_string()))),
        "calls": agg(&calls),
        "puts": agg(&puts),
        "put_call_ratio": ((total_put_vol as f64 / (total_call_vol.max(1)) as f64) * 10000.0).round() / 10000.0,
        "total_call_volume": total_call_vol,
        "total_put_volume": total_put_vol,
        "unusual": unusual,
    }))
}
